otherlibs/systhreads/st_win32.h

```diff
@@ -158,30 +158,88 @@ Caml_inline void st_thread_yield(st_masterlock * m)
   EnterCriticalSection(m);
 }
 
+/* List of threads waiting on a mutex or on a condition variable.
+   Each thread is represented by its associated event. */
+
+struct st_wait_list {
+  HANDLE event;                  /* event of the first waiting thread */
+  struct st_wait_list * next;
+};
+
+/* Recover (or create) the event associated with the calling thread */
+static HANDLE st_event_for_current_thread(void)
+{
+  HANDLE ev = (HANDLE) TlsGetValue(st_thread_sem_key);
+  if (ev == 0) {
+    ev = CreateEvent(NULL,
+                     FALSE /*auto reset*/,
+                     FALSE /*initially unset*/,
+                     NULL);
+    if (ev == 0) return 0;
+    TlsSetValue(st_thread_sem_key, (void *) ev);
+  }
+  return ev;
+}
+
 /* Mutexes */
 
-typedef CRITICAL_SECTION * st_mutex;
+struct st_mutex_struct {
+  CRITICAL_SECTION lock;        /* protect the data structure */
+  enum { UNLOCKED, LOCKED } status;
+  struct st_wait_list * waiters; /* list of threads waiting to lock it */
```

Comment:
I think this is more heavyweight than it needs to be - a single manual reset event should work for this, as only broadcast semantics are needed - `SetEvent` and `ResetEvent` would only be called with the `lock`.

As an aside, I _think_ it's more heavyweight than is needed for condition variables as well - I think `st_condvar_struct` could simply have two Event objects, one auto-reset and one manual-reset - `st_condvar_wait` would then do a `WaitForMultipleObjects` with `bWaitAll == FALSE` and you'd use the auto-reset event for `st_condvar_signal` (wakes up just one thread and resets the event) and the manual-reset event for `st_condvar_broadcast` (wakes them all up), but that's O/T for this PR.